File: dotnet/CardsProject/Domain/Tournaments/TournamentRegistration.cs
using CardsProject.Domain.Players;
using CardsProject.Domain.Cards;
using System.ComponentModel.DataAnnotations.Schema;

namespace CardsProject.Domain.Tournaments;

public enum TournamentRegistrationStatusType
{
    Registered,
    Waitlisted,
    Withdrawn,
    Disqualified
}

public class TournamentRegistration
{
    public int Id { get; set; }

    public TournamentRegistrationStatusType Status { get; set; }
    public int? Seed { get; set; } = null;
    public int? FinalStanding { get; set; } = null;
    public int PointsEarned { get; set; } = 0;
    public DateTime? RegisteredAt { get; set; } = null;

    public int? TournamentId { get; set; }
    [ForeignKey(nameof(TournamentId))]
    public Tournament? Tournament { get; set; }
    public int? PlayerId { get; set; }
    [ForeignKey(nameof(PlayerId))]
    public Player? Player { get; set; }
    public int? DeckId { get; set; }
    [ForeignKey(nameof(DeckId))]
    public Deck? Deck { get; set; }

    // Business operations

    public void Withdraw()
    {
        throw new NotImplementedException("withdraw not implemented");
    }

    public void Disqualify(string reason)
    {
        throw new NotImplementedException("disqualify not implemented");
    }

    public void PromoteFromWaitlist()
    {
        throw new NotImplementedException("promote_from_waitlist not implemented");
    }
}

File: phoenix/lib/cards_project/tournaments/tournament_registration.ex
defmodule CardsProject.Tournaments.TournamentRegistration do
  use Ecto.Schema
  import Ecto.Changeset

  schema "tournament_registrations" do
    field :status, :string
    field :seed, :integer
    field :final_standing, :integer
    field :points_earned, :integer, default: 0
    field :registered_at, :naive_datetime
    belongs_to :tournament, CardsProject.Tournaments.Tournament
    belongs_to :player, CardsProject.Players.Player
    belongs_to :deck, CardsProject.Cards.Deck

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:points_earned, :registered_at, :status, :seed, :final_standing, :tournament_id, :player_id, :deck_id])
    |> validate_required([:points_earned, :registered_at])
    |> validate_inclusion(:status, ["Registered", "Waitlisted", "Withdrawn", "Disqualified"])
  end

  # ── Business operations ────────────────────────────────────────────

  def withdraw(_record) do
    # TODO: implement TournamentRegistration.withdraw
    :ok
  end

  def disqualify(_record, _reason) do
    # TODO: implement TournamentRegistration.disqualify
    :ok
  end

  def promote_from_waitlist(_record) do
    # TODO: implement TournamentRegistration.promote_from_waitlist
    :ok
  end
end
